### PR TITLE
Remove `should_rank_save_checkpoint` property from TTP

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Finishes #9074

### Does your PR introduce any breaking changes? If yes, please list them.

TTP API (experimental): removes the property

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [n/a] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda @kaushikb11 @justusschock @awaelchli @akihironitta @ananthsub

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 1]
```diff
@@ -312,7 +312,16 @@ def save_checkpoint(self, checkpoint: Dict[str, Any], filepath: _PATH) -> None:
...
-        return self.checkpoint_io.save_checkpoint(checkpoint, filepath)
+        self.checkpoint_io.save_checkpoint(checkpoint, filepath)
+
+    def remove_checkpoint(self, filepath: _PATH) -> None:
+        """Remove checkpoint filepath from the filesystem.
+
+        Args:
+            filepath: Path to checkpoint
+        """
+        if self.local_rank == 0:
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:323]: This change is equivalent to what's in master. However, I don't understand why we don't check local rank for `save_checkpoint` but we do it for remove.

File: pytorch_lightning/plugins/training_type/single_tpu.py [hunk 2]
```diff
@@ -82,7 +82,7 @@ def save_checkpoint(self, checkpoint: Dict[str, Any], filepath: _PATH) -> None:
             checkpoint: dict containing model and trainer state
             filepath: write-target file's path
         """
-        return self.checkpoint_io.save_checkpoint(checkpoint, filepath)
+        self.checkpoint_io.save_checkpoint(checkpoint, filepath)
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/single_tpu.py:85]: Can't we remove this override? Isn't global rank always 0 for single tpu?

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:323]: because xm.save needs to be called on all ranks. It internally checks if the local rank is 0. This is just how the library works. If we don't call xm.save on all ranks, it will hang. 

https://pytorch.org/xla/release/1.5/_modules/torch_xla/core/xla_model.html

File: pytorch_lightning/plugins/training_type/single_tpu.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/single_tpu.py:85]: it was overridden because of my answer below.

Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/single_tpu.py:85]: I think you can drop the override

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:323]: Okay, I'll add a comment

File: pytorch_lightning/plugins/training_type/single_tpu.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/plugins/training_type/single_tpu.py]: Done

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 3]
```diff
@@ -312,7 +312,17 @@ def save_checkpoint(self, checkpoint: Dict[str, Any], filepath: _PATH) -> None:
             checkpoint: dict containing model and trainer state
             filepath: write-target file's path
         """
-        return self.checkpoint_io.save_checkpoint(checkpoint, filepath)
```
Comment (four4fish) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:315]: save_checkpoint() is same as training_type_plugin, should we remove it?

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 4]
```diff
@@ -312,7 +312,17 @@ def save_checkpoint(self, checkpoint: Dict[str, Any], filepath: _PATH) -> None:
...
+        # xm.save needs to be called on all ranks. It internally checks if the local rank is 0
+        self.checkpoint_io.save_checkpoint(checkpoint, filepath)
+
+    def remove_checkpoint(self, filepath: _PATH) -> None:
```
Comment (four4fish) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:318]: RFC: should we have checkpoint_io for tpu and deepspeed instead of override functions in strategies directly?

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 3 above)
Comment (jjenniferdai) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:315]: rank check is different (default global_zero v.s. all ranks here)

**Comment by jjenniferdai** [comment]
@carmocca @ananthsub see https://github.com/PyTorchLightning/pytorch-lightning/pull/11068#discussion_r769903035 re: Trainer property removal v.s. deprecation?

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 3 above)
Comment (four4fish) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:315]: @jjenniferdai the logic difference is in XLACheckpointIO.save_checkpoint() right? the strategy overrides here is same as training_type_plugin so it's not needed?

Comment (jjenniferdai) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:315]: yep IO differences in `checkpointIO` but there's also different rank logic controlled by TTPs hence the TTP override:

training_type_plugin: `if global_zero: checkpointIO.save()`
here: `checkpointIO.save()` (all ranks)

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 4 above)
Comment (jjenniferdai) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:318]: TPU: override is just rank logic which should be controlled by strategies right

deepspeed: my thoughts are it would be hard/inadvisable to fit inside `IOPlugin`, since it looks like it involves things like `broadcast` call, checking trainer state, which we ideally shouldn't do in `IOPlugin`. I think we'd need to pass `deepspeed_engine` too
https://github.com/PyTorchLightning/pytorch-lightning/blob/3c4d06bd42ff3282a017cd9fa879bbd68e8568df/pytorch_lightning/plugins/training_type/deepspeed.py#L755-L768

**Comment by carmocca** [comment]
@jjenniferdai We don't need to keep BC for the property in the TTP. Since the property in the `Trainer` is read-only, we can copy its current implementation into it.

Whichever PR lands later should be updated.

File: pytorch_lightning/trainer/trainer.py [hunk 5]
```diff
@@ -1702,9 +1703,10 @@ def world_size(self) -> int:
     @property
     def should_rank_save_checkpoint(self) -> bool:
         rank_zero_deprecation(
-            "`Trainer.should_rank_save_checkpoint` is deprecated in v1.6 and will be removed in 1.8.", stacklevel=5
+            "`Trainer.should_rank_save_checkpoint` is deprecated in v1.6 and will be removed in v1.8.", stacklevel=5
         )
-        return self.training_type_plugin.should_rank_save_checkpoint
+        ttp = self.training_type_plugin
+        return isinstance(ttp, TPUSpawnPlugin) and ttp.local_rank == 0 or ttp.is_global_zero
```
Comment (jjenniferdai) [inline pytorch_lightning/trainer/trainer.py:1709]: re: custom plugins like @awaelchli 's comment [here](https://github.com/PyTorchLightning/pytorch-lightning/pull/11068#pullrequestreview-833560843):
could check for `hasattr(should_rank_save_checkpoint)` but could be misleading when TTP's `should_rank_save_checkpoint` is removed and doesn't actually control anything but the trainer property is not yet removed

Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:1709]: I'd say this is okay since it is read-only. IMO the deprecation here is more about not explicitly breaking the property access.

I also wouldn't worry about custom plugins since we are absolutely not recommending them for now. And when we do (1.6+), users won't implement `should_rank_save_checkpoint` as this PR will have removed it.

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 6]
```diff
@@ -331,10 +341,6 @@ def all_gather(self, tensor: torch.Tensor, group: Optional[Any] = None, sync_gra
     def teardown(self) -> None:
         os.environ.pop("PT_XLA_DEBUG", None)
 
-    @property
-    def should_rank_save_checkpoint(self) -> bool:
-        return self.local_rank == 0
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:336]: I think this was answered here: https://github.com/PyTorchLightning/pytorch-lightning/pull/7296#discussion_r631789547

[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:336]: Yes exactly right. We wouldn't have this property in the first place if it wasn't for this difference. #7684

File: pytorch_lightning/trainer/trainer.py (hunk 5 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:1709]: I still think if we are not going all the way, we should remove the Trainer property without deprecation, or raise a custom error message. After all, it was just meant to be a short hand for `trainer.training_type_plugin.xyz`

Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:1709]: I'm totally fine either way. I proposed marking the property protected from the start
https://github.com/PyTorchLightning/pytorch-lightning/pull/7684#discussion_r639336673

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 7]
```diff
@@ -312,7 +312,17 @@ def save_checkpoint(self, checkpoint: Dict[str, Any], filepath: _PATH) -> None:
...
+        # `xla_model.save` needs to be called on all ranks. It internally checks if the local rank is 0
+        self.checkpoint_io.save_checkpoint(checkpoint, filepath)
+
+    def remove_checkpoint(self, filepath: _PATH) -> None:
+        """Remove checkpoint filepath from the filesystem.
+
+        Args:
+            filepath: Path to checkpoint
+        """
+        if self.local_rank == 0:
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:324]: This could be an issue in the future if the nodes filesystem becomes one a single volume.

Ideally, I think we should detect if the filesystem is shared. The condition would become (if self.global_rank if shared_filesystem else self.local_rank)

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:336]: Instead we could have:

```py
save_rank = self.is_global_zero if self.checkpoint_io.shared_filesystem else self.local_rank
if save_rank:
    ...
```

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 7 above)
Comment (carmocca) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:324]: I believe you have the same question as this comment: https://github.com/PyTorchLightning/pytorch-lightning/pull/11070/files#r770116930

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:324]: Yes, mind-opening an issue associated with this?

File: pytorch_lightning/trainer/trainer.py (hunk 5 above)
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:1717]: Keeping the deprecation since nobody seems to have a strong opinion and the work has been done already

**Review by Borda — APPROVED** [review]
